ocaml/xapi-idl/cluster/cluster_interface.ml

```diff
@@ -15,12 +15,51 @@ type debug_info = string [@@deriving rpcty]
 (** Name of the cluster *)
 type cluster_name = string [@@deriving rpcty]
 
-(** An IPv4 address (a.b.c.d) *)
-type address = IPv4 of string [@@deriving rpcty]
+type ip = IPv4 of string | IPv6 of string [@@deriving rpcty]
 
-let printaddr () = function IPv4 s -> Printf.sprintf "IPv4(%s)" s
+let string_of_ip = function IPv4 a | IPv6 a -> a
 
-let str_of_address address = match address with IPv4 a -> a
+(** this address includes the hostname and hostuuid along with an ip address
+  this is done to maintain backwards compatability, and should be combined with
+  the other variant in the future.
+  *)
+type extended_addr = {ip: ip; hostuuid: string; hostname: string}
+[@@deriving rpcty]
+
+type address = IPv4 of string | Extended of extended_addr [@@deriving rpcty]
+
+let ipstr_of_address = function
+  | IPv4 a ->
+      a
+  | Extended {ip; _} -> (
+    match ip with IPv4 a | IPv6 a -> a
+  )
+
+let ipaddr_of_address = function
+  | IPv4 _ as ip ->
+      ip
+  | Extended _ as extended_addr ->
+      IPv4 (ipstr_of_address extended_addr)
```

Comment:
This can't be right when IPv6 is used

Reply:
I don't think we implemented IPV6 for clustering, we probably should.

Reply:
Yeah, it looks like the old code was all exclusively IPv4 and this PR adds a little bit, but not complete, IPv6 stuff.

Reply:
not the intent of this PR to add complete IPv6 support, just some initial PoC. But good point here to make it explicit (and maybe add a comment)